network/src/commonMain/kotlin/com/wire/kalium/network/api/user/details/UserDetailsApiImp.kt

```diff
@@ -1,20 +1,83 @@
 package com.wire.kalium.network.api.user.details
 
+import com.wire.kalium.network.api.ErrorResponse
+import com.wire.kalium.network.exceptions.KaliumException
 import com.wire.kalium.network.utils.NetworkResponse
+import com.wire.kalium.network.utils.isSuccessful
+import com.wire.kalium.network.utils.mapSuccess
 import com.wire.kalium.network.utils.wrapKaliumResponse
 import io.ktor.client.HttpClient
 import io.ktor.client.request.post
 import io.ktor.client.request.setBody
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.coroutineScope
 
 class UserDetailsApiImp(private val httpClient: HttpClient) : UserDetailsApi {
 
-    override suspend fun getMultipleUsers(users: ListUserRequest): NetworkResponse<List<UserDetailsResponse>> = wrapKaliumResponse {
-        httpClient.post(PATH_LIST_USERS) {
-            setBody(users)
+    override suspend fun getMultipleUsers(users: ListUserRequest): NetworkResponse<List<UserDetailsResponse>> {
+        val chunkedRequests = separateRequestInChunks(users)
+        val results = performChunkedRequests(chunkedRequests)
+        if (results.isEmpty()){
+            return NetworkResponse.Error(
+                KaliumException.InvalidRequestError(
+                    ErrorResponse(
+                        400,
+                        "No values passed. At least one userId or handle should be passed in order to get user details",
+                        "missing-values"
+                    )
+                )
+            )
+        }
+
+        // If any request fails, abort and return the failure
+        results.firstOrNull { !it.isSuccessful() }?.let { errorResponse ->
+            return errorResponse as NetworkResponse.Error
+        }
+
+        // Join all the successful results together into a single one
+        return results.fold(results.first()) { acc, item ->
+            item as NetworkResponse.Success
+            acc.mapSuccess { currentValue ->
+                currentValue + item.value
+            }
+        }
+    }
+
+    private suspend fun performChunkedRequests(chunkedRequests: List<ListUserRequest>): List<NetworkResponse<List<UserDetailsResponse>>> =
+        coroutineScope {
```

Comment:
Just dumb question, why do need to spcify this `coroutineScope` before the api call? doesn't it come implicitly within the suspended function bounds?

Reply:
The problem is the `Collections.map{}` call.
It doesn't take a `suspend` lambda.

This way, we can open a `coroutineScope`, perform the async requests in parallel and wait for them to complete side-by-side. It's actually faster this way too 🏎️ 🚀